### DISCUSSION TITLE
[native-image] @Deleted classes

### DISCUSSION THREAD

**Discussion by Vinche59** [oracle/graal#3656] [Q&A] [answered]
Hello,

I use an old library who heavily use `java.util.Calendar$Builder.build` making `java.util.JapaneseImperialCalendar` type reacheable during compilation and this class is marked @Deleted by GraalVM in [CalendarSubstitutions.java](https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/CalendarSubstitutions.java)

I'd want to know if there's a mechanism to "Undelete" a @Deleted class when building a native image with GraalVM (without using `--report-unsupported-elements-at-runtime` option) ?

**Comment by zakkak** [comment, marked as answer]
HI @Vinche59, my understanding is that those are marked as `@deleted` because they are essentially not supported. Even if there was a way to "Undelete" them your application would probably fail if they ever got reached, that said using `--report-unsupported-elements-at-runtime` is probably a safer way to handle this, since it will throw an exception if the corresponding classes get reached.

Unfortunately though `--report-unsupported-elements-at-runtime` is discouraged in production code, and also makes it harder to detect other issues in your `native-image`.

**Reply by Vinche59** [reply]
Hello @zakkak 
Thanks for the answer.

Yes it's what I was afraid of... So I will refactor and try to drop that library.

Just for info (as I see you quite active in Quarkus project also), I don't succeed to compile with `--report-unsupported-elements-at-runtime` option under Quarkus. As soon as I add this option, I've got compile errors in native (mostly `NoClassDefFound` errors) and as you said this option is discourage in production code :-)

**Reply by zakkak** [reply]
@Vinche59 that sounds a lot like https://github.com/oracle/graal/issues/1725
So the work around would be to manually add the missing dependencies like in https://github.com/quarkusio/quarkus/issues/6327#issuecomment-789483700

As you can see it's a bit of a slippery slope so if you can avoid the said library it would be best :)

Please open an issue at https://github.com/quarkusio/quarkus or start a discuss at quarkusio.zulipchat.com/ if you need more help.
